control-plane/csi-driver/src/bin/node/format.rs

```diff
@@ -1,46 +1,66 @@
 //! Utility function for formatting a device with filesystem
-use csi_driver::context::FileSystem;
+use crate::mount;
+use csi_driver::filesystem::FileSystem;
 
-use devinfo::blkid::probe::Probe;
-use std::process::Command;
-use tracing::{debug, trace};
+use tracing::debug;
+use uuid::Uuid;
 
-pub(crate) async fn prepare_device(device: &str, fstype: &FileSystem) -> Result<(), String> {
+/// Prepare the filesystem before mount, change parameters if requested.
+pub(crate) async fn prepare_device(
+    fstype: &FileSystem,
+    volume_uuid: &Uuid,
+    device: &str,
+    staging_path: &str,
+    options: &[String],
+    change_fs_id: bool,
+) -> Result<(), String> {
     debug!("Probing device {}", device);
+    let fs = FileSystem::get_property(device, "TYPE");
 
-    let probe =
-        Probe::new_from_filename(device).map_err(|error| format!("probe setup failed: {error}"))?;
+    let fs_ops = fstype.to_fs_ops()?;
 
-    if let Err(error) = probe.do_probe() {
-        return Err(format!("probe failed: {error}"));
-    }
-
-    if let Ok(fs) = probe.lookup_value("TYPE") {
+    if let Ok(fs) = fs {
         debug!("Found existing filesystem ({}) on device {}", fs, device);
+        if change_fs_id {
+            match fstype {
+                FileSystem::Ext4 => {
+                    debug!("Attempting repair for {volume_uuid}, {fstype} filesystem");
+                    fs_ops.set_uuid(device, &volume_uuid.to_string()).await?;
+                }
+                FileSystem::Xfs => {
+                    debug!("Attempting mounting and unmounting for {volume_uuid}, to repair {fstype} filesystem");
+                    repair_xfs_filesystem(device, staging_path, options, fstype, volume_uuid)?;
+                    fs_ops.set_uuid(device, &volume_uuid.to_string()).await?;
+                }
+                _ => {}
+            }
+            debug!("Changed {fstype} filesystem uuid to {volume_uuid} for {device}");
+        }
+
         return Ok(());
     }
 
     debug!("Creating new filesystem ({}) on device {}", fstype, device);
+    fs_ops.create(device).await
+}
 
-    let binary = format!("mkfs.{fstype}");
-    let output = Command::new(&binary)
-        .arg(device)
-        .output()
-        .map_err(|error| format!("failed to execute {binary}: {error}"))?;
-
-    trace!(
-        "Output from {} command: {}",
-        binary,
-        String::from_utf8(output.stdout.clone()).unwrap()
-    );
-
-    if output.status.success() {
-        return Ok(());
-    }
-
-    Err(format!(
-        "{} command failed: {}",
-        binary,
-        String::from_utf8(output.stderr).unwrap()
-    ))
+/// Xfs filesystem needs an unmount to clear the log, so that the parameters can be changed.
+/// Mount the filesystem to a defined path and then unmount it.
+fn repair_xfs_filesystem(
```

Comment:
Shouldn't this be in the trait?

Reply:
I was not able to put the xfs_repair in the trait because the mount module is not accessible in the FileSystem lib.